File: typescript/packages/core/src/vfs/bound.ts
import type { Accessor } from '../accessor/base.ts'
import { resolveGlobOf, type CommandIO } from '../commands/builtin/generic_bind/adapter.ts'
import { PathSpec, type FileStat } from '../types.ts'
import { mountKey, mountPrefixOf } from '../utils/key_prefix.ts'
import { BaseVFS, type FindOptions } from './base.ts'
import { VFSAdapter } from './adapter.ts'

/** Shared binding for builtin and custom VFS; lifecycle belongs to the subclass. */

export abstract class BoundVFS<A extends Accessor = Accessor> extends BaseVFS {
  declare writeFile?: (path: PathSpec, data: Uint8Array) => Promise<void>

  declare appendFile?: (path: PathSpec, data: Uint8Array) => Promise<void>

  declare exists?: (path: PathSpec) => Promise<boolean>

  declare mkdir?: (path: PathSpec, options?: { recursive?: boolean }) => Promise<void>

  declare rmdir?: (path: PathSpec) => Promise<void>

  declare unlink?: (path: PathSpec) => Promise<void>

  declare rename?: (src: PathSpec, dst: PathSpec) => Promise<void>

  declare truncate?: (path: PathSpec, length: number) => Promise<void>

  declare copy?: (src: PathSpec, dst: PathSpec) => Promise<void>

  declare rmR?: (path: PathSpec) => Promise<void>

  declare du?: (path: PathSpec) => Promise<number>

  declare find?: (path: PathSpec, options?: FindOptions) => Promise<string[]>

  abstract readonly accessor: A
  readonly io: CommandIO<A>

  constructor(table: CommandIO<A> | VFSAdapter<A>) {
    super()
    this.io = table instanceof VFSAdapter ? table.toCommandIO() : table
    this.#installOptional(this.io)
  }
  // Each forwarder reads `this.index` when called rather than capturing
  // it, because `setIndex` can replace it after construction.
  #installOptional(io: CommandIO<A>): void {
    const { write, append, exists, mkdir, rmdir, unlink } = io
    const { rename, truncate, copy, rmR, du, find } = io
    if (write !== undefined && this.writeFile === undefined)
      this.writeFile = (p, d) => write(this.accessor, p, d)
    if (append !== undefined && this.appendFile === undefined)
      this.appendFile = (p, d) => append(this.accessor, p, d)
    if (exists !== undefined && this.exists === undefined)
      this.exists = (p) => exists(this.accessor, p)
    if (mkdir !== undefined && this.mkdir === undefined)
      this.mkdir = (p, o) => mkdir(this.accessor, p, o?.recursive)
    if (rmdir !== undefined && this.rmdir === undefined) this.rmdir = (p) => rmdir(this.accessor, p)
    if (unlink !== undefined && this.unlink === undefined)
      this.unlink = (p) => unlink(this.accessor, p)
    if (rename !== undefined && this.rename === undefined)
      this.rename = (s, d) => rename(this.accessor, s, d)
    if (truncate !== undefined && this.truncate === undefined)
      this.truncate = (p, n) => truncate(this.accessor, p, n)
    if (copy !== undefined && this.copy === undefined)
      this.copy = (s, d) => copy(this.accessor, s, d)
    if (rmR !== undefined && this.rmR === undefined) this.rmR = (p) => rmR(this.accessor, p)
    if (du !== undefined && this.du === undefined)
      this.du = (p) => du.size(this.accessor, p, this.index)
    if (find !== undefined && this.find === undefined)
      this.find = (p, o) => find(this.accessor, p, o ?? {})
  }

  glob(paths: readonly PathSpec[], prefix = ''): Promise<PathSpec[]> {
    const effective = prefix
      ? paths.map((p) =>
          mountPrefixOf(p.virtual, p.vfsPath)
            ? p
            : new PathSpec({
                virtual: p.virtual,
                directory: p.directory,
                ...(p.pattern !== null ? { pattern: p.pattern } : {}),
                resolved: p.resolved,
                vfsPath: mountKey(p.virtual, prefix),
              }),
        )
      : paths
    return resolveGlobOf(this.io)(this.accessor, effective, this.index)
  }
  readFile(path: PathSpec): Promise<Uint8Array> {
    return this.io.readBytes(this.accessor, path, this.index)
  }

  readdir(path: PathSpec): Promise<string[]> {
    return this.io.readdir(this.accessor, path, this.index)
  }

  stat(path: PathSpec): Promise<FileStat> {
    return this.io.stat(this.accessor, path, this.index)
  }

  streamPath(path: PathSpec): AsyncIterable<Uint8Array> {
    return this.io.readStream(this.accessor, path, this.index)
  }
}

File: python/mirage/vfs/bound.py
from collections.abc import Awaitable, Callable
from dataclasses import replace
from typing import Any

from mirage.cache.index import IndexCacheStore, IndexConfig
from mirage.commands.builtin.generic_bind.adapter import CommandIO
from mirage.types import PathSpec
from mirage.utils.key_prefix import mount_key
from mirage.vfs.adapter import VFSAdapter
from mirage.vfs.base import BaseVFS

_DIRECT_OPS: dict[str, str] = {
    "readdir": "readdir",
    "read_bytes": "read_bytes",
    "range_read": "read_range",
    "read_stream": "read_stream",
    "stat": "stat",
    "write": "write",
    "append": "append",
    "create": "create",
    "mkdir": "mkdir",
    "unlink": "unlink",
    "rmdir": "rmdir",
    "rm_recursive": "rm_r",
    "rename": "rename",
    "copy": "copy",
    "truncate": "truncate",
    "exists": "exists",
    "find_flat": "find",
}


def range_adapter(
    fn: Callable[..., Awaitable[bytes]],
    index: Callable[[], IndexCacheStore],
) -> Callable[..., Awaitable[bytes]]:
    """Adapt a table's ``read_range`` slot to the VFS API's shape.

    The one field whose two callers disagree about more than a name. A
    builtin publishes ``range_read(accessor, path, start, end)`` with the
    end exclusive, while the table slot is ``(accessor, path, index,
    offset, size)``, and ``BaseVFS.__getattr__`` binds only the
    accessor. Forwarded raw, the table function took ``start`` as its
    index and ``end`` as its offset, so an object-store table read from
    ``end`` to EOF instead of the window asked for, and one that touched
    the index at all crashed on an int.

    The index is read per call rather than captured, because
    ``set_index`` can replace the store after construction.

    Args:
        fn (Callable): the table's ``read_range`` slot.
        index (Callable): reads the owning VFS's current index.
    """

    async def range_read(accessor: Any, path: PathSpec, start: int,
                         end: int) -> bytes:
        return await fn(accessor, path, index(), start, end - start)

    return range_read


def direct_ops(
    io: CommandIO,
    index: Callable[[], IndexCacheStore],
) -> dict[str, Callable[..., Any]]:
    """Derive the direct filesystem surface from the operation table.

    Args:
        io (CommandIO): callbacks shared with commands and filesystem ops.
        index (Callable): supplies the current index for range adaptation.
    """
    ops: dict[str, Callable[..., Any]] = {}
    for name, field in _DIRECT_OPS.items():
        fn = getattr(io, field)
        if fn is None:
            continue
        ops[name] = (range_adapter(fn, index) if field == "read_range" else fn)
    if io.du is not None:
        ops["du_size"] = io.du.size
        ops["du_entries"] = io.du.entries
    return ops


class BoundVFS(BaseVFS):
    """Bind direct filesystem calls to the same table as commands and ops."""

    def __init__(self,
                 *,
                 io: CommandIO | VFSAdapter,
                 index: IndexConfig | None = None) -> None:
        super().__init__(index=index)
        self.io = io.to_command_io() if isinstance(io, VFSAdapter) else io
        self._ops = direct_ops(self.io, lambda: self.index)

    async def resolve_glob(self,
                           paths: list[PathSpec],
                           prefix: str = "") -> list[PathSpec]:
        if prefix:
            paths = [
                replace(p, vfs_path=mount_key(p.virtual, prefix))
                for p in paths
            ]
        return await self.io.resolve_glob(self.accessor, paths, self.index)
